Tiltfile

```diff
@@ -396,6 +399,120 @@ def prepare_all():
     )
     local(cmd, env = settings.get("kustomize_substitutions", {}))
 
+# create flavor resources from cluster-template files in the templates directory
+def flavors():
+    substitutions = settings.get("kustomize_substitutions", {})
+
+    # TODO: should we set `./test/infrastructure/docker/templates`as the default?
+    flavor_dirs = settings.get("flavor_dirs", [])
+
+    for flavor_dir in flavor_dirs:
+        template_list = [item for item in listdir(flavor_dir)]
+        template_list = [template_name for template_name in template_list if os.path.basename(template_name).endswith("yaml")]
+
+        for template_name in template_list:
+            deploy_templates(template_name, substitutions)
+
+def deploy_templates(template_name, substitutions):
+    # validate template_name exists
+    if not os.path.exists(template_name):
+        fail(template_name + " not found")
+
+    yaml = str(read_file(template_name))
+    yaml_stream = read_yaml_stream(template_name)
+    print(yaml_stream)
+
+    # TODO: document these substitutions that are used in the default templates
+    # NAMESPACE
+    # KUBERNETES_VERSION
+    # CONTROL_PLANE_MACHINE_COUNT
+    # WORKER_MACHINE_COUNT
+
+    for substitution in substitutions:
+        value = substitutions[substitution]
+        yaml = yaml.replace("${" + substitution + "}", value)
+
+    yaml = yaml.replace('"', '\\"')  # add escape character to double quotes in yaml
+
+    basename = os.path.basename(template_name)
```

Comment:
Can we please explore if there is any way that we can just reuse `clusterctl generate cluster` instead of reimplementing it. (with `--form` arg)

Reply:
This should also take care of deploying the ClusterClass of a template automatically.

Reply:
Sure I'm can look into that. Could you clarify the use case here for `clusterctl generate --form`? It looks like given one of our Cluster templates, it will fill in the cluster name, namespace, worker machine count, ctrl plane machine count, and k8s version. However, I'm not sure how it's supposed to work for ClusterClasses. It looks like it outputs another ClusterClass template when `clusterclass-quick-start.yaml` is passed in.

Reply:
It also seems like we'd be getting the variable substitutions from the shell and wouldn't be able to use `kustomize_substitutions` in that case. Any thoughts @fabriziopandini @CecileRobertMichon?

Reply:
We should only use `clusterctl generate cluster` on cluster-templates (not ClusterClasses). If the cluster-template uses a ClusterClass it will be automatically rendered and then deployed too.

I think we should be able to set the kustomize_substitutions as env vars when running `clusterctl generate cluster` (we do the same for envsubst on this PR today)

My main motivation is to try to avoid reimplementing clustectl generate cluster and it's logic regarding substitution, ClusterClass lookup, ...

Reply:
@sbueringer Could you clarify a bit more about how to use `clusterctl generate cluster` with a cluster class? Maybe I'm misunderstanding but I don't think the clusterclass is automatically applied. I followed [these steps](https://cluster-api.sigs.k8s.io/user/quick-start.html#generating-the-cluster-configuration) with `--from cluster-template-development-topology.yaml`, and I got an error saying that the clusterclass isn't found. 

```
Error from server (InternalError): error when creating "capi-quickstart.yaml": admission webhook "default.cluster.cluster.x-k8s.io" denied the request: Internal error occurred: Cluster test-cluster can't be validated. ClusterClass quick-start can not be retrieved: ClusterClass.cluster.x-k8s.io "quick-start" not found
```

Reply:
I think that's due to the `--from`. When used with regular releases it works automatically.

@ykakarap WDYT is the best way to also use a cluster-template with ClusterClass with --from? (e.g. either another parameter or to configure a local repository)

Reply:
Okay gotcha. Since the template directory had a ClusterClass and a cluster template using that ClusterClass, I figured the workflow was to apply the class and then apply the template. But if we can find a way to accomplish this with only one file that would be ideal. Does that mean we want try to use clusterctl to apply the class automatically when the template is applied, or to try to automatically generate the template given the class?

Reply:
> Does that mean we want try to use clusterctl to apply the class automatically when the template is applied, or to try to automatically generate the template given the class?

We would run `clusterctl generate cluster` and the output would include the Cluster and the ClusterClass automatically, which we then can deploy with clusterctl.

Let's see what Yuvaraj says if it's already possible today, or if there's a way to make that possible.

Reply:
@ykakarap If you have some time, your input would be greatly appreciated :)

Reply:
Clarifying a few things so that we are all on the same page: `clusterctl generate cluster` always starts with `cluster-template-{flavor}.yaml`. If the cluster template referenced by the flavour uses a ClusterClass then clusterctl checks if the ClusterClass is already installed in the API server, if not it adds the ClusterClass definition (available in file `clusterclass-{class-name}.yaml`) to the template and outputs the final result.

The above flow is not supported when using `--from`. The `--from` flow is very simple in what it does: take the file form the location, substitute variables and output it.

There was previously a discussion about supporting ClusterClass flow with `--from` under this issue: https://github.com/kubernetes-sigs/cluster-api/issues/5496. TL;DR of the discussion: The conclusion was to not try to support it as using a local repository would be a better solution in general but let's come back to it if there is a good use case.

I think we are all on agreement that we dont want to re-implement the `clusterctl generate cluster` logic in the Tiltfile and re-use clusterctl. With that we have 3 possible approaches we can take here:

1) Spin up a local repository during `tilt up` that holds the template and ClusterClass definitions and use that for `clusterctl generate cluster`. This will work with both classic and classy clusters. 
    * Note that since the management cluster is not initialized with `clusterctl` the inventory will not be populated and we should not rely on auto-discovery for any of the values. All the provider information and versions will have to be passed explicitly.
    * To support creating local repository we can restructure `flavor_dirs` to something like:

```yaml=
flavor_dirs:
  docker: <path-to-docker-templates> # includes the clusterclass definitions
  azure: <path-to-azure-templates> # includes the clusterclass definitions
```

2) Re-investigate adding ClusterClass support to `--from` flow and use that.
    * The issues stated in the https://github.com/kubernetes-sigs/cluster-api/issues/5496 still hold for this usecase and we should also consider the implications of adding the support becasue once we add it we cannot stop users from using it in other places outside the Tilt setup and that would increase the UX surface we will have to support.

3) Install all found ClusterClass definitions during `tilt up`. Then users can simply deploy any flavour (clusterclass based or otherwise).
    * In this approach we dont have to use `clusterctl` and the deployment flow could simply be just applying the target template.
    * In this approach we will have to wait for the webhooks to come up before we install the ClusterClass definitions during `tilt up`.

Reply:
I think I would prefer 1 or 2

With 3. we have the problems that:
* we always have to deploy all providers for which we have cluster templates locally
* not sure if the names of all objects of all ClusterClasses are unique (especially across providers)
* I would prefer only deploying a ClusterClass if we need it, otherwise we have a lot of resources deployed that we don't need

I also see 1 or 2 as a chance to test clusterctl implicitly with our tilt env

Reply:
>not sure if the names of all objects of all ClusterClasses are unique (especially across providers)

Good point!

>I think I would prefer 1 or 2

+1 for option 1 over 2. 

>I also see 1 or 2 as a chance to test clusterctl implicitly with our tilt env

Agreed. For the initial implementation I guess we will only use clusterctl for `generate cluster`. In later iterations, if it ties well with Tilt, we can probably also initialize the cluster using clusterctl and do other operations with it.

Reply:
`clusterctl` uses a 'repository' as the source of files that it needs to perform various operations like `init`, `generate`, etc. In this case the repository will be used to look for the Cluster templates and the ClusterClass definitions. `clusterctl` defaults to looking for these files under github releases. However we can create a local file system based repository with the necessary files needed and have clusterctl point to that to look for the files it cares about. 

There is a hack script in CAPI repo that we use to create a local repository: https://github.com/kubernetes-sigs/cluster-api/blob/4d7ec52225855f3c13179656f6358e80319ab5e8/cmd/clusterctl/hack/create-local-repository.py. (This script probably does a lot more than we need in this use case - might be useful for referene). Additional notes about creating and using a local repository can be found in the CAPI book: https://cluster-api.sigs.k8s.io/clusterctl/developers.html#create-the-local-repository

For this feature we will probably want to spin up a local repository with just the Cluster templates and the ClusterClass definitions (we need only these for `clusterctl generate` to work) from the information we get in tilt-setting.yaml and then use that when running the `clusterctl generate` command so that we can handle both classic and classy clusters similarly.

Reply:
@ykakarap Okay gotcha. If I understand this correctly, we need to add the path to our templates to `clusterctl-settings.json` in the `provider_repos` field.